File: tests/test_effects.py
import inspect

import pytest

from cougar_rgb.device import LED_COUNT, Fusion2
from cougar_rgb.effects import EFFECTS, IDLE_CHOICES, Params, apply_brightness, parse_color, to_hex


def params_for(effect, **kwargs):
    return Params(colors=effect.default_colors or ['#ffffff'], name=effect.name, **kwargs)


@pytest.mark.parametrize('name', list(EFFECTS))
@pytest.mark.parametrize('speed', [1, 5, 10])
def test_render_produces_valid_frames(name, speed):
    effect = EFFECTS[name]
    p = params_for(effect, speed=speed, random=effect.has_random)
    for t in (0.0, 0.016, 0.5, 3.3, 100.0):
        frame = apply_brightness(effect.render(t, p), 80)
        assert len(frame) == LED_COUNT
        assert all(0 <= c <= 255 for rgb in frame for c in rgb)


@pytest.mark.parametrize('name', [n for n, e in EFFECTS.items() if e.kind == 'hw'])
def test_hw_packet_matches_set_effect_signature(name):
    effect = EFFECTS[name]
    kwargs = effect.hw_packet(params_for(effect))
    inspect.signature(Fusion2.set_effect).bind(None, 5, **kwargs)
    assert 0 <= kwargs.get('max_brightness', 255) <= 255


def test_hw_brightness_uses_full_byte_range():
    static = EFFECTS['static']
    assert static.hw_packet(Params(colors=['#ff0000'], brightness=100))['max_brightness'] == 255
    assert static.hw_packet(Params(colors=['#ff0000'], brightness=0))['max_brightness'] == 0


def test_hw_breathing_speed_matches_rgb_fusion_ranges():
    pulse = EFFECTS['breathing']
    fastest = pulse.hw_packet(Params(colors=['#ffffff'], speed=10))['periods'][0]
    slowest = pulse.hw_packet(Params(colors=['#ffffff'], speed=1))['periods'][0]
    assert (fastest, slowest) == (400, 1600)


def test_apply_brightness_extremes():
    frame = [(255, 128, 1)] * LED_COUNT
    assert apply_brightness(frame, 100) == frame
    assert apply_brightness(frame, 0) == [(0, 0, 0)] * LED_COUNT


def test_color_helpers_roundtrip():
    assert parse_color('#1a2b3c') == (0x1A, 0x2B, 0x3C)
    assert to_hex((0x1A, 0x2B, 0x3C)) == '#1a2b3c'


def test_idle_choices_are_plain_software_effects():
    assert IDLE_CHOICES[0] == 'none'
    for name in IDLE_CHOICES[1:]:
        assert EFFECTS[name].kind == 'sw' and not EFFECTS[name].audio


def test_audio_effects_preview_without_analyzer():
    for effect in (e for e in EFFECTS.values() if e.audio):
        p = params_for(effect)
        frames = [effect.render(i / 60, p) for i in range(120)]
        assert len({tuple(f) for f in frames}) > 1, effect.name

File: tests/test_effects.cpp
#include "effects.hpp"
#include "fakes.hpp"

#include <catch2/catch_test_macros.hpp>
#include <catch2/generators/catch_generators.hpp>
#include <set>

using namespace cougar;

namespace {

Params params_for(const Effect &e, int speed = 5)
{
    Params p;
    p.name = e.name();
    p.colors.clear();
    for (const auto &c : e.info().default_colors)
        p.colors.push_back(*parse_color(c));
    if (p.colors.empty())
        p.colors.push_back({255, 255, 255});
    p.speed = speed;
    p.random = e.info().has_random;
    return p;
}

}  // namespace

TEST_CASE("every effect renders valid frames")
{
    test::TestClock clock;
    const int speed = GENERATE(1, 5, 10);
    for (const auto &e : effects()) {
        const Params p = params_for(*e, speed);
        for (double t : {0.0, 0.016, 0.5, 3.3, 100.0}) {
            const RenderFrame frame = e->render(t, p);
            for (const Rgb &c : frame) {
                INFO(e->name() << " t=" << t);
                CHECK((c.r >= -1e-9 && c.r <= 255 + 1e-9 && c.g >= -1e-9 && c.g <= 255 + 1e-9 &&
                       c.b >= -1e-9 && c.b <= 255 + 1e-9));
            }
        }
    }
}

TEST_CASE("hardware brightness uses the full byte range")
{
    const Effect &e = *find_effect("static");
    Params p = params_for(e);
    p.brightness = 100;
    CHECK(e.hw_packet(p).max_brightness == 255);
    p.brightness = 0;
    CHECK(e.hw_packet(p).max_brightness == 0);
}

TEST_CASE("hardware breathing speed matches RGB Fusion ranges")
{
    const Effect &e = *find_effect("breathing");
    CHECK(e.hw_packet(params_for(e, 10)).periods[0] == 400);
    CHECK(e.hw_packet(params_for(e, 1)).periods[0] == 1600);
}

TEST_CASE("hardware effect colors are passed through")
{
    const Effect &e = *find_effect("static");
    Params p = params_for(e);
    p.colors = {{255, 16, 1}};
    CHECK(e.hw_packet(p).color == Rgb8{255, 16, 1});
}

TEST_CASE("apply_brightness extremes")
{
    RenderFrame frame;
    frame.fill({255, 128, 1});
    Frame full;
    full.fill({255, 128, 1});
    CHECK(apply_brightness(frame, 100) == full);
    CHECK(apply_brightness(frame, 0) == Frame{});
}

TEST_CASE("color helpers roundtrip")
{
    const auto c = parse_color("#1a2b3c");
    REQUIRE(c);
    CHECK((c->r == 0x1A && c->g == 0x2B && c->b == 0x3C));
    CHECK(to_hex(*c) == "#1a2b3c");
    CHECK(parse_color("1A2B3C"));
    CHECK_FALSE(parse_color("#12345"));
    CHECK_FALSE(parse_color("#12345g"));
}

TEST_CASE("idle choices are plain software effects")
{
    const auto &choices = idle_choices();
    CHECK(choices.front() == IDLE_NONE);
    for (size_t i = 1; i < choices.size(); ++i) {
        const Effect *e = find_effect(choices[i]);
        REQUIRE(e);
        CHECK((!e->hardware() && !e->info().audio));
    }
}

TEST_CASE("audio effects preview a simulated beat without the analyzer")
{
    for (const auto &e : effects()) {
        if (!e->info().audio)
            continue;
        const Params p = params_for(*e);
        std::set<std::tuple<int, int, int>> distinct;
        for (int i = 0; i < 120; ++i) {
            const Frame f = apply_brightness(e->render(i / 60.0, p), 100);
            distinct.insert({f[0].r, f[0].g, f[0].b});
        }
        INFO(e->name());
        CHECK(distinct.size() > 1);
    }
}
